NeoML/src/TraditionalML/WordDictionary.cpp

```diff
@@ -0,0 +1,205 @@
+/* Copyright � 2017-2022 ABBYY Production LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+--------------------------------------------------------------------------------------------------------------*/
+
+#include <common.h>
+#pragma hdrstop
+
+#include <NeoML/TraditionalML/WordDictionary.h>
+
+namespace NeoML {
+
+bool CWordDictionary::CWordWithCount::operator<( const CWordDictionary::CWordWithCount& other ) const
+{
+	if( Count < other.Count ) {
+		return true;
+	} else if( Count == other.Count ) {
+		return Word > other.Word;
+	} else {
+		return false;
+	}
+}
+
+bool CWordDictionary::CWordWithCount::operator==( const CWordDictionary::CWordWithCount& other ) const
+{
+	return Count == other.Count
+		&& Word == other.Word;
+}
+
+void CWordDictionary::CWordWithCount::Serialize( CArchive& archive )
+{
+	archive.Serialize( Word );
+	archive.Serialize( Count );
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////
+
+CWordDictionary::CWordDictionary( CWordDictionary&& other ) :
+	totalWordsUseCount( other.totalWordsUseCount )
+{
+	other.words.MoveTo( words );
+	other.wordToId.MoveTo( wordToId );
+}
+
+CWordDictionary& CWordDictionary::operator=( CWordDictionary&& other )
+{
+	totalWordsUseCount = other.totalWordsUseCount;
+	other.words.MoveTo( words );
+	other.wordToId.MoveTo( wordToId );
+	return *this;
+}
+
+void CWordDictionary::CopyTo( CWordDictionary& other ) const
+{
+	other.totalWordsUseCount = totalWordsUseCount;
+	words.CopyTo( other.words );
+	wordToId.CopyTo( other.wordToId );
+}
+
+void CWordDictionary::AddWord( const CString& word, long long addedCount )
+{
+	int id = 0;
+	if( !wordToId.Lookup( word, id ) ) {
+		wordToId.Set( word, words.Size() );
+		words.Add( CWordWithCount{ word, addedCount } );
+	} else {
+		words[id].Count += addedCount;

```

Comment:
_optional_ `NeoAssert( words[id].Count> 0 )`?

Reply:
When we introduce new token during BPE training algorithm we need to decrease counts of some pairs, some of them can become zero-counted.
Added a softer check for non-negativity.